{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This notebook aims to re-calculate strand bias using the three methods described by [Guo2012](https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-13-666). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We start from the three trimmed.sorted.bam files listed below. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first three commands below use \"scripts/freqs_modified.py\", which generate variants tables containing: \n",
    "- VariantCov\n",
    "- ForwardVariantCov\n",
    "- ReverseVariantCov\n",
    "- RefCov\n",
    "- ForwardRefCov\n",
    "- ReverseRefCov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python scripts/freqs_modified.py --snpfreqmin 0.03 bam/BC01.trimmed.sorted.bam refs/ZIKV_REF.fasta > BC01_modified.variants.0.03.txt\n",
    "!python scripts/freqs_modified.py --snpfreqmin 0.03 bam/BC02.trimmed.sorted.bam refs/ZIKV_REF.fasta > BC02_modified.variants.0.03.txt\n",
    "!python scripts/freqs_modified.py --snpfreqmin 0.03 bam/BC03.trimmed.sorted.bam refs/ZIKV_REF.fasta > BC03_modified.variants.0.03.txt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The three commands below are trying to reproduce the generation of BC01/02/03.variants.0.03.txt from BC01/02/03.trimmed.sorted.bam using \"scripts/freqs.py\". BC01/02/03.trimmed.sorted.bam and scripts/freqs.py were taken from [Nicholas J. Loman](https://github.com/nickloman/zika-isnv/).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python scripts/freqs.py --snpfreqmin 0.03 bam/BC01.trimmed.sorted.bam refs/ZIKV_REF.fasta > BC01_confirm.variants.0.03.txt\n",
    "!python scripts/freqs.py --snpfreqmin 0.03 bam/BC02.trimmed.sorted.bam refs/ZIKV_REF.fasta > BC02_confirm.variants.0.03.txt\n",
    "!python scripts/freqs.py --snpfreqmin 0.03 bam/BC03.trimmed.sorted.bam refs/ZIKV_REF.fasta > BC03_confirm.variants.0.03.txt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Conclusion: BC01/02/03_modified.variants.0.03.txt and BC01/02/03_confirm.variants.0.03.txt are the same, except that BC01/02/03_modified.variants.0.03.txt contain three more columns—RefCov, ForwardRefCov, and ReverseRefCov. \n",
    "#### BC01/02/03_confirm.variants.0.03.txt (generated here) are different compared to BC01/02/03.variants.0.03.txt (from [Nicholas J. Loman](https://github.com/nickloman/zika-isnv/)), the commands and input files used are exactly the same though. BC01/02/03.trimmed.sorted.bam by [Nicholas J. Loman](https://github.com/nickloman/zika-isnv/) should be updated.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: ── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.0 ──\n",
      "\n",
      "R[write to console]: \u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.2     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.0.2     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.0     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.3.1     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.0\n",
      "\n",
      "R[write to console]: ── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R \n",
    "library(\"tidyverse\")\n",
    "options(dplyr.width = Inf)\n",
    "options(readr.num_columns = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "# Strand bias calculation, formula taken from Guo2012.  \n",
    "SB_calc = function(ForwardRefCov, ForwardVariantCov, ReverseRefCov, ReverseVariantCov){\n",
    "    a = ForwardRefCov\n",
    "    b = ForwardVariantCov\n",
    "    c = ReverseRefCov\n",
    "    d = ReverseVariantCov\n",
    "    bias = abs(b/(a+b) - d/(c+d)) / ((b+d)/(a+b+c+d))\n",
    "    return (bias)\n",
    "}\n",
    "\n",
    "# GATK strand bias calculation, formula taken from Guo2012. \n",
    "GATK_SB_calc = function(ForwardRefCov, ForwardVariantCov, ReverseRefCov, ReverseVariantCov){\n",
    "    a = ForwardRefCov\n",
    "    b = ForwardVariantCov\n",
    "    c = ReverseRefCov\n",
    "    d = ReverseVariantCov\n",
    "    bias1 = b/(a+b)*(c/(c+d))/((a+c)/(a+b+c+d))\n",
    "    bias2 = d/(c+d)*(a/(a+b))/((a+c)/(a+b+c+d))\n",
    "    return(pmax(bias1,bias2))\n",
    "}\n",
    "\n",
    "# Fisher strand bias calculation, formula taken from Guo2012. \n",
    "Fisher_SB_calc = function(ForwardRefCov, ForwardVariantCov, ReverseRefCov, ReverseVariantCov){\n",
    "    a = ForwardRefCov\n",
    "    b = ForwardVariantCov\n",
    "    c = ReverseRefCov\n",
    "    d = ReverseVariantCov\n",
    "    testor = matrix(c(a,b,c,d), ncol=2)\n",
    "    return (1 - fisher.test(testor)$p.value)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90m# A tibble: 3 x 8\u001b[39m\n",
      "\u001b[90m# Rowwise: \u001b[39m\n",
      "       Pos ForwardVariantCov ReverseVariantCov ForwardRefCov ReverseRefCov    SB\n",
      "     \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m         \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m         \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m 43\u001b[4m9\u001b[24m\u001b[4m1\u001b[24m\u001b[4m7\u001b[24m013                 2                 0            11            20  2.54\n",
      "\u001b[90m2\u001b[39m 95\u001b[4m9\u001b[24m\u001b[4m2\u001b[24m\u001b[4m3\u001b[24m670                 2                 0            16            10  1.56\n",
      "\u001b[90m3\u001b[39m 57\u001b[4m0\u001b[24m\u001b[4m8\u001b[24m\u001b[4m8\u001b[24m850                 2                 0             8            16  2.6 \n",
      "  GATK_SB Fisher_SB\n",
      "    \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m   0.164     0.852\n",
      "\u001b[90m2\u001b[39m   0.120     0.476\n",
      "\u001b[90m3\u001b[39m   0.217     0.862\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# Test the three strand calculation methods on Guo2012 Table 3 \n",
    "# Filter out variants with ForwardRefCov <= 10 or ForwardVariantCov <= 10 or ReverseRefCov <= 10 or ReverseVariantCov <= 10. \n",
    "guo = read_tsv(\"Guo_table_3.txt\")\n",
    "guo %>%\n",
    "  mutate(SB = SB_calc(ForwardRefCov, ForwardVariantCov, ReverseRefCov, ReverseVariantCov)) %>%\n",
    "  mutate(GATK_SB = GATK_SB_calc(ForwardRefCov, ForwardVariantCov, ReverseRefCov, ReverseVariantCov)) %>% \n",
    "  rowwise() %>% \n",
    "  mutate(Fisher_SB = Fisher_SB_calc(ForwardRefCov, ForwardVariantCov, ReverseRefCov, ReverseVariantCov))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SB, GATK_SB, and Fisher_SB are the same compared to those provided in table 3 in Guo2012. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next is to apply clustering models on the MinION data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%R \n",
    "# Read in the three datasets \"BC01.variants.0.03.txt\", \"BC02.variants.0.03.txt\", \"BC03.variants.0.03.txt\". \n",
    "barcode1snps=read_tsv(\"BC01_modified.variants.0.03.txt\")\n",
    "barcode1snps$replica = 'a'\n",
    "\n",
    "barcode2snps=read_tsv(\"BC02_modified.variants.0.03.txt\")\n",
    "barcode2snps$replica = 'b'\n",
    "\n",
    "barcode3snps=read_tsv(\"BC03_modified.variants.0.03.txt\")\n",
    "barcode3snps$replica = 'c'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# Merge the three datasets\n",
    "minion_variants=rbind(barcode1snps, barcode2snps, barcode3snps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove position in amplicons with primer mismatches.\n",
    "\n",
    "These positions are described in \"expectedpositions.txt\", where each of the position are labelled by: \n",
    "- \"REMOVE\": positions to be removed, because they are located in the amplicons with primer mismatches\n",
    "- \"TRUE\": true positives variants by comparing the consensus nt sequence of the gb records KX087101 and KU955593. \n",
    "- \"FALSE\": false positives, variants detected by MinION, but not labelled as \"TRUE\". "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "expectedpositions=read_tsv(\"expectedpositions.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "# Remove positions with the label \"REMOVE\"\n",
    "minion_variants_positions=minion_variants %>%\n",
    "   left_join(expectedpositions, by=c(\"Pos\" = \"Position\")) %>%\n",
    "   filter(State != 'Remove')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# Add strand bias info\n",
    "# Note that here we do not filter out position \n",
    "# with (ForwardRefCov or ForwardVariantCov or ReverseRefCov or ReverseVariantCov < 10)\n",
    "# Instead, we filter out rows with NaN in SB or GATK_SB or Fisher_SB. \n",
    "\n",
    "minion_variants_positions %>%\n",
    "  mutate(Decision = ifelse(grepl(\"TRUE\", State), 1, 0)) %>%\n",
    "  mutate(SB = SB_calc(ForwardRefCov, ForwardVariantCov, ReverseRefCov, ReverseVariantCov)) %>%\n",
    "  mutate(GATK_SB = GATK_SB_calc(ForwardRefCov, ForwardVariantCov, ReverseRefCov, ReverseVariantCov)) %>% \n",
    "  rowwise() %>% \n",
    "  mutate(Fisher_SB = Fisher_SB_calc(ForwardRefCov, ForwardVariantCov, ReverseRefCov, ReverseVariantCov)) %>% \n",
    "  filter(!is.na(SB)) %>%\n",
    "  filter(!is.na(GATK_SB)) %>%\n",
    "  filter(!is.na(Fisher_SB)) %>%\n",
    "  write_tsv(\"minion_variants_positions_SB.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Codes below are written by me in Python. These codes construct \n",
    "- a logistic regression classifier using Freq and StrandAF, \n",
    "- a KNN classifier using Freq and StrandAF, \n",
    "- a SVM classifier using Freq and StrandAF. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "import pandas as pd \n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.linear_model import LogisticRegressionCV\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import metrics\n",
    "from sklearn.metrics import accuracy_score\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "from itertools import cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pos</th>\n",
       "      <th>Qual</th>\n",
       "      <th>Freq</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Base</th>\n",
       "      <th>UngappedCoverage</th>\n",
       "      <th>TotalCoverage</th>\n",
       "      <th>VariantCov</th>\n",
       "      <th>ForwardVariantCov</th>\n",
       "      <th>ReverseVariantCov</th>\n",
       "      <th>RefCov</th>\n",
       "      <th>ForwardRefCov</th>\n",
       "      <th>ReverseRefCov</th>\n",
       "      <th>replica</th>\n",
       "      <th>State</th>\n",
       "      <th>Decision</th>\n",
       "      <th>SB</th>\n",
       "      <th>GATK_SB</th>\n",
       "      <th>Fisher_SB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1070</td>\n",
       "      <td>0</td>\n",
       "      <td>0.091876</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1034</td>\n",
       "      <td>1035</td>\n",
       "      <td>95</td>\n",
       "      <td>49</td>\n",
       "      <td>46</td>\n",
       "      <td>938</td>\n",
       "      <td>558</td>\n",
       "      <td>380</td>\n",
       "      <td>a</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.296377</td>\n",
       "      <td>0.109318</td>\n",
       "      <td>0.845000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1129</td>\n",
       "      <td>0</td>\n",
       "      <td>0.038532</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>545</td>\n",
       "      <td>548</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>522</td>\n",
       "      <td>134</td>\n",
       "      <td>388</td>\n",
       "      <td>a</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>3.078417</td>\n",
       "      <td>0.128516</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1175</td>\n",
       "      <td>0</td>\n",
       "      <td>0.247379</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>477</td>\n",
       "      <td>515</td>\n",
       "      <td>118</td>\n",
       "      <td>105</td>\n",
       "      <td>13</td>\n",
       "      <td>359</td>\n",
       "      <td>55</td>\n",
       "      <td>304</td>\n",
       "      <td>a</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>2.487032</td>\n",
       "      <td>0.836195</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1176</td>\n",
       "      <td>0</td>\n",
       "      <td>0.043544</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>666</td>\n",
       "      <td>668</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>637</td>\n",
       "      <td>178</td>\n",
       "      <td>459</td>\n",
       "      <td>a</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>3.217391</td>\n",
       "      <td>0.146475</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1183</td>\n",
       "      <td>0</td>\n",
       "      <td>0.031088</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>386</td>\n",
       "      <td>390</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>366</td>\n",
       "      <td>365</td>\n",
       "      <td>1</td>\n",
       "      <td>a</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>27.658537</td>\n",
       "      <td>0.908081</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>10342</td>\n",
       "      <td>0</td>\n",
       "      <td>0.030395</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>329</td>\n",
       "      <td>333</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>317</td>\n",
       "      <td>223</td>\n",
       "      <td>94</td>\n",
       "      <td>c</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1.403433</td>\n",
       "      <td>0.044272</td>\n",
       "      <td>0.931737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>10349</td>\n",
       "      <td>0</td>\n",
       "      <td>0.131959</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>485</td>\n",
       "      <td>487</td>\n",
       "      <td>64</td>\n",
       "      <td>36</td>\n",
       "      <td>28</td>\n",
       "      <td>419</td>\n",
       "      <td>152</td>\n",
       "      <td>267</td>\n",
       "      <td>c</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.728833</td>\n",
       "      <td>0.199787</td>\n",
       "      <td>0.996427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>10360</td>\n",
       "      <td>0</td>\n",
       "      <td>0.059671</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>486</td>\n",
       "      <td>487</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>457</td>\n",
       "      <td>277</td>\n",
       "      <td>180</td>\n",
       "      <td>c</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2.195685</td>\n",
       "      <td>0.142643</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>10361</td>\n",
       "      <td>0</td>\n",
       "      <td>0.121348</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>445</td>\n",
       "      <td>465</td>\n",
       "      <td>54</td>\n",
       "      <td>26</td>\n",
       "      <td>28</td>\n",
       "      <td>391</td>\n",
       "      <td>249</td>\n",
       "      <td>142</td>\n",
       "      <td>c</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.578174</td>\n",
       "      <td>0.169730</td>\n",
       "      <td>0.964145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>10376</td>\n",
       "      <td>0</td>\n",
       "      <td>0.107345</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>531</td>\n",
       "      <td>531</td>\n",
       "      <td>57</td>\n",
       "      <td>33</td>\n",
       "      <td>24</td>\n",
       "      <td>473</td>\n",
       "      <td>387</td>\n",
       "      <td>86</td>\n",
       "      <td>c</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1.298132</td>\n",
       "      <td>0.225266</td>\n",
       "      <td>0.999904</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>435 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Pos  Qual      Freq Ref Base  UngappedCoverage  TotalCoverage  \\\n",
       "0     1070     0  0.091876   C    T              1034           1035   \n",
       "1     1129     0  0.038532   G    A               545            548   \n",
       "2     1175     0  0.247379   A    G               477            515   \n",
       "3     1176     0  0.043544   G    A               666            668   \n",
       "4     1183     0  0.031088   C    T               386            390   \n",
       "..     ...   ...       ...  ..  ...               ...            ...   \n",
       "430  10342     0  0.030395   A    T               329            333   \n",
       "431  10349     0  0.131959   T    C               485            487   \n",
       "432  10360     0  0.059671   C    T               486            487   \n",
       "433  10361     0  0.121348   T    C               445            465   \n",
       "434  10376     0  0.107345   G    A               531            531   \n",
       "\n",
       "     VariantCov  ForwardVariantCov  ReverseVariantCov  RefCov  ForwardRefCov  \\\n",
       "0            95                 49                 46     938            558   \n",
       "1            21                 19                  2     522            134   \n",
       "2           118                105                 13     359             55   \n",
       "3            29                 29                  0     637            178   \n",
       "4            12                  4                  8     366            365   \n",
       "..          ...                ...                ...     ...            ...   \n",
       "430          10                 10                  0     317            223   \n",
       "431          64                 36                 28     419            152   \n",
       "432          29                  1                 28     457            277   \n",
       "433          54                 26                 28     391            249   \n",
       "434          57                 33                 24     473            387   \n",
       "\n",
       "     ReverseRefCov replica  State  Decision         SB   GATK_SB  Fisher_SB  \n",
       "0              380       a   True         1   0.296377  0.109318   0.845000  \n",
       "1              388       a  False         0   3.078417  0.128516   1.000000  \n",
       "2              304       a   True         1   2.487032  0.836195   1.000000  \n",
       "3              459       a  False         0   3.217391  0.146475   1.000000  \n",
       "4                1       a  False         0  27.658537  0.908081   1.000000  \n",
       "..             ...     ...    ...       ...        ...       ...        ...  \n",
       "430             94       c  False         0   1.403433  0.044272   0.931737  \n",
       "431            267       c   True         1   0.728833  0.199787   0.996427  \n",
       "432            180       c  False         0   2.195685  0.142643   1.000000  \n",
       "433            142       c   True         1   0.578174  0.169730   0.964145  \n",
       "434             86       c   True         1   1.298132  0.225266   0.999904  \n",
       "\n",
       "[435 rows x 19 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"minion_variants_positions_SB.tsv\", sep=\"\\t\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
